sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbStateStoreProvider.scala

```diff
@@ -0,0 +1,642 @@
+          } else {
+            val path = getCheckpointPath(version)
+            val r: RocksDbInstance =
+              new RocksDbInstance(keySchema, valueSchema, version.toString, rocksDbConf)
+            r.open(path, readOnly = true)
+            r.iterator(closeDbOnCompletion = true)
+          }
+        case COMMITTED =>
+          logDebug(s"state = committed using check-pointed DB with version $newVersion")
+          // use check-pointed db for current updated version
+          val path = getCheckpointPath(newVersion)
+          val r: RocksDbInstance =
+            new RocksDbInstance(keySchema, valueSchema, newVersion.toString, rocksDbConf)
+          r.open(path, readOnly = true)
+          r.iterator(closeDbOnCompletion = true)
+
+        case _ => Iterator.empty
+      }
+    }
+
+    override def metrics: StateStoreMetrics = {
+      val metricsFromProvider: Map[String, Long] = getMetricsForProvider()
+      val customMetrics = metricsFromProvider.flatMap {
+        case (name, value) =>
+          // just allow searching from list cause the list is small enough
+          supportedCustomMetrics.find(_.name == name).map(_ -> value)
+      }
+      StateStoreMetrics(Math.max(numEntriesInDb, 0), Math.max(bytesUsedByDb, 0), customMetrics)
+    }
+
+    /*
+     * Whether all updates have been committed
+     */
+    override def hasCommitted: Boolean = {
+      state == COMMITTED
+    }
+
+    override def toString(): String = {
+      s"RocksDbStateStore[id=(op=${id.operatorId},part=${id.partitionId}),dir=$baseDir]"
+    }
+
+  }
+
+  /*
+   * Initialize the provider with more contextual information from the SQL operator.
+   * This method will be called first after creating an instance of the StateStoreProvider by
+   * reflection.
+   *
+   * @param stateStoreId    Id of the versioned StateStores that this provider will generate
+   * @param keySchema       Schema of keys to be stored
+   * @param valueSchema     Schema of value to be stored
+   * @param keyIndexOrdinal Optional column (represent as the ordinal of the field in keySchema) by
+   *                        which the StateStore implementation could index the data.
+   * @param storeConfs      Configurations used by the StateStores
+   * @param hadoopConf      Hadoop configuration that could be used by StateStore
+   *                        to save state data
+   */
+  override def init(
+      stateStoreId: StateStoreId,
+      keySchema: StructType,
+      valueSchema: StructType,
+      keyIndexOrdinal: Option[Int], // for sorting the data by their keys
+      storeConfs: StateStoreConf,
+      hadoopConf: Configuration): Unit = {
+    this.stateStoreId_ = stateStoreId
+    this.keySchema = keySchema
+    this.valueSchema = valueSchema
+    this.storeConf = storeConfs
+    this.hadoopConf = hadoopConf
+    this.numberOfVersionsToRetain = storeConfs.maxVersionsToRetainInMemory
+    fm.mkdirs(baseDir)
+    this.rocksDbConf = storeConf.confs
```

Comment:
Let me explain my point. If we take an example like `RocksDbInstance.scala` it contains implementation detail about how to get specific parameter values, like [this](https://github.com/apache/spark/pull/24922/files#diff-1d0920add5494799c03eed9682fe7bfbR199-R203). In this scenario it's bad that `RocksDbInstance` defines what is the default etc. This can be extracted to `RocksDbStateStoreConf` class which can inherit all the behavior from `StateStoreConf`. If it's applied then the implementation would look this this: `conf.blockSizeInKB`. As mentioned in my previous comment `ConfigBuilder` can be used and together with this change the code would be less coupled. Something like this:

Possibility1.
```
class RocksDbStateStoreConf(@transient private val stateStoreConf: StateStoreConf)
  extends StateStoreConf {
...
  val blockSizeInKB: Int = stateStoreConf.sqlConf.blockSizeInKB
...
}
```

Possibility2.
```
class RocksDbStateStoreConf(@transient private val sqlConf: SQLConf)
  extends StateStoreConf {
...
  val blockSizeInKB: Int = sqlConf.blockSizeInKB
...
}
```
Haven't spent too much time which one is better but I've the feeling the first one is easier. Feel free to evaluate and suggest better option if you've something.

Reply:
Sure. Let me make the changes.

Reply:
Went ahead with Possibility 1. 

sqlConf is private in stateStoreConf. So made the following [changes](https://github.com/apache/spark/pull/24922/files#diff-094eeec6658534cb1c29a11d4a5c0f5e) in this PR

if we are okay to make `sqlConf` public, we can have this [change](https://github.com/itsvikramagr/spark/blob/SPARK-28120-sqlconfchanges/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbStateStoreConf.scala) for RocksDbConf and corresponding changes in stateStoreConf and sqlConf